### PR TITLE
Add empty `pin` command

### PR DESCRIPTION
**Pull request by florelis**
This change adds an empty `pin` command. Actual implementation of the command will come as a separate PR. This change includes:
* Definition of the `pin` command and its sub-commands `add`, `remove`, `list`, `reset`. These definitions contain only the definition of arguments but are missing the actual execution code.
* A new experimental feature flag to hide this command.
* All the help resource strings for the command. Still pending all the resources that will be shown during actual execution.
* Adds a `source ls` alias for the `source list` command.
* Changes an existing string to make it consistent with the rest.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2733)

### CODE REVIEW

**Comment by florelis** [comment]
Related to #476 and #2611 but does not resolve them yet

File: src/AppInstallerCLICore/Commands/PinCommand.cpp [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+#include "Resources.h"
+
+namespace AppInstaller::CLI
+{
+    using namespace AppInstaller::CLI::Execution;
+    using namespace AppInstaller::Utility::literals;
+    using namespace std::string_view_literals;
+
+    // TODO: Create this link with the docs!
+    static constexpr std::string_view s_PinCommand_HelpLink = "https://aka.ms/winget-command-pin"sv;
```
Comment (florelis) [inline src/AppInstallerCLICore/Commands/PinCommand.cpp:17]: @denelon Can you create this even if it points to just the general documentation (or add a note for you to do that)? I remember with a previous command I added it we completely forgot about it...

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 2]
```diff
@@ -1357,7 +1357,8 @@ Please specify one of them using the `--source` option to proceed.</value>
     <value>Portable install failed; Cleaning up...</value>
```
Comment (florelis) [inline src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:1357]: I really want to sort this file alphabetically but I know VS is just going to mess it up again...

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw [hunk 3]
```diff
@@ -1483,4 +1484,47 @@ Please specify one of them using the `--source` option to proceed.</value>
...
+    <comment>{Locked{"'upgrade'"} Locked{"--include-pinned"} Locked{"winget upgrade --all"}</comment>
+  </data>
+  <data name="PinAddCommandShortDescription" xml:space="preserve">
+    <value>Add a new pin</value>
+  </data>
+  <data name="PinCommandLongDescription" xml:space="preserve">
+    <value>Manage package pins with the sub-commands. A pin can limit the Windows Package Manager from updating a package to specific ranges of versions, or it can prevent it from updating the package altogether. A pinned package may still update on its own and be updated from outside the Windows Package Manager.</value>
+  </data>
+  <data name="PinCommandShortDescription" xml:space="preserve">
+    <!-- TODO: We need to add "package pin[ning]" to the dictionary of terms for the loc team to know before sending this for localization -->
```
Comment (florelis) [inline src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:1498]: Have we already done this for the other terms we use, like "source" or "package"?

File: src/AppInstallerCommonCore/Public/winget/ExperimentalFeature.h [hunk 4]
```diff
@@ -25,6 +25,7 @@ namespace AppInstaller::Settings
             DirectMSI = 0x2,
             ZipInstall = 0x4,
             OpenLogsArgument = 0x8,
+            Pinning = 0x10,
```
Comment (florelis) [inline src/AppInstallerCommonCore/Public/winget/ExperimentalFeature.h:28]: It took me longer than I want to admit to remember what 2 x 8 is in hex 😅

File: src/AppInstallerCLICore/Commands/SourceCommand.h [hunk 5]
```diff
@@ -37,7 +37,7 @@ namespace AppInstaller::CLI
 
     struct SourceListCommand final : public Command
     {
-        SourceListCommand(std::string_view parent) : Command("list", parent) {}
+        SourceListCommand(std::string_view parent) : Command("list", { "ls" }, parent) {}
```
Comment (florelis) [inline src/AppInstallerCLICore/Commands/SourceCommand.h:40]: This came to mind while I was adding the alias for the `pin` sub-commands. I don't know if we want to make this a separate PR so it goes out sooner than pinning.

File: src/AppInstallerCLICore/Commands/PinCommand.cpp (hunk 1 above)
Comment (denelon) [reply src/AppInstallerCLICore/Commands/PinCommand.cpp:17]: If I remember correctly, aka.ms will just default somewhere else if the URL isn't registered.
Yes, we can create the aka.ms links ahead of time with a temporary/placeholder URL and modify it at any point later.

**Comment by denelon** [comment]
This should also include the update to the settings page as an experimental feature.

File: doc/Settings.md [hunk 6]
```diff
@@ -262,3 +263,14 @@ Experimental feature with the aim of managing dependencies, as of now it only sh
...
 ```
+
+### pinning
+
+This feature enables the ability to pin packages to a specific version to prevent the Windows Package Manager from updating them.
+You can enable the feature as shown below.
+
+```json
+   "experimentalFeatures": {
+       "openLogsArgument": true
```
Comment (Trenly) [inline doc/Settings.md:274]:
```suggestion
       "pinning": true
```

File: src/AppInstallerCLICore/Commands/PinCommand.cpp [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+    std::vector<std::unique_ptr<Command>> PinCommand::GetCommands() const
+    {
+        return InitializeFromMoveOnly<std::vector<std::unique_ptr<Command>>>({
+            std::make_unique<PinAddCommand>(FullName()),
+            std::make_unique<PinRemoveCommand>(FullName()),
+            std::make_unique<PinListCommand>(FullName()),
```
Comment (Trenly) [inline src/AppInstallerCLICore/Commands/PinCommand.cpp:24]: Does `reset` need to be included here?

File: src/AppInstallerCLICore/Commands/PinCommand.cpp (hunk 1 above)
Comment (Trenly) [reply src/AppInstallerCLICore/Commands/PinCommand.cpp:17]: Aka.ms links that don't exist just redirect to the Bing homepage iirc

File: src/AppInstallerCLICore/Commands/SourceCommand.h (hunk 5 above)
Comment (Trenly) [reply src/AppInstallerCLICore/Commands/SourceCommand.h:40]: I would suggest a separate PR just to keep the git history cleaner

File: src/AppInstallerCLICore/Commands/PinCommand.cpp (hunk 1 above)
Comment (florelis) [reply src/AppInstallerCLICore/Commands/PinCommand.cpp:17]: Yeah, they redirect to the Bing homepage. I was mentioning creating it now pointing to the general documentation, because if we do forget about creating the right link, in my mind it's better if the link at least points to somewhere close instead of going to Bing.

File: src/AppInstallerCLICore/Commands/SourceCommand.h (hunk 5 above)
Comment (florelis) [reply src/AppInstallerCLICore/Commands/SourceCommand.h:40]: Yeah, you're right... I have the bad habit of seeing things that I want to change but then being too lazy to do it in a separate PR...

Anyways, here it is #2736

File: doc/Settings.md (hunk 6 above)
Comment (florelis) [reply doc/Settings.md]: Right. That's what I get for just copying the one above 😅

File: src/AppInstallerCLICore/Commands/PinCommand.cpp (hunk 7 above)
Comment (florelis) [reply src/AppInstallerCLICore/Commands/PinCommand.cpp:24]: Added!

**Comment by Trenly** [comment]
@jsoref

**Comment by jsoref** [comment]
@Trenly: interesting, when I tested it, I got a very different result: https://github.com/check-spelling/winget-cli/pull/2#issuecomment-1336272299 (this comment has since been collapsed, but you can expand it).

From looking at the logs ([here](https://github.com/microsoft/winget-cli/actions/runs/3605526073/jobs/6076011084#step:2:508) and [there](https://github.com/check-spelling/winget-cli/actions/runs/3610555666/jobs/6084422806#step:2:508)), I suspect that [checkout-merge step](https://github.com/check-spelling/check-spelling/blob/d7cd2973c513e84354f9d6cf50a6417a628a78ce/action.yml#L324-L326) is not doing the right thing tm.

Here'a a rebased run:

https://github.com/check-spelling/winget-cli/pull/2/files#file-src-powershell-microsoft-winget-client-examples-sample_findpackage-ps1-L4

And here's me _trying_ to reproduce this PR as it was:
https://github.com/check-spelling/winget-cli/pull/3

I'm going to have to investigate the checkout-merge logic, it feels like it isn't quite behaving.

Can I ask that this be rebased after #2739 is merged? (Preferably squashing 74a97e6 into 1ede614f7b8a168ddbc30e7454aed24be5898a21 -- when I did a rebase, I tripped on a conflict there...)

File: src/AppInstallerCLICore/Commands/PinCommand.cpp (hunk 1 above)
Comment (denelon) [reply src/AppInstallerCLICore/Commands/PinCommand.cpp:17]: I added https://aka.ms/winget-command-pin

It currently defaults to the main winget page.

**Comment by florelis** [comment]
Thanks for looking into this @jsoref !

> Can I ask that this be rebased after #2739 is merged? 
Rebased. Looks like it found one unrecognized word but I'm not planning on merging it with that line so I'll leave it like this for now.

File: src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw (hunk 3 above)
Comment (florelis) [reply src/AppInstallerCLIPackage/Shared/Strings/en-us/winget.resw:1499]: Looks like not. I added pinning to the internal work item tracking these.

File: doc/Settings.md [hunk 8]
```diff
@@ -262,3 +263,14 @@ Experimental feature with the aim of managing dependencies, as of now it only sh
        "dependencies": true
    },
 ```
+
+### pinning
+
+This feature enables the ability to pin packages to a specific version to prevent the Windows Package Manager from updating them.
```
Comment (yao-msft) [inline doc/Settings.md:269]: nit: it does not have to be a specific version

File: src/AppInstallerCLICore/Commands/SourceCommand.h [hunk 9]
```diff
@@ -37,7 +37,7 @@ namespace AppInstaller::CLI
 
     struct SourceListCommand final : public Command
     {
-        SourceListCommand(std::string_view parent) : Command("list", { "ls" }, parent) {}
+        SourceListCommand(std::string_view parent) : Command("list", parent) {}
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Commands/SourceCommand.h:40]: Will this actually remove the "ls" change done in previous pr during merging?

File: src/AppInstallerCLICore/Commands/PinCommand.cpp [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+            Argument::ForType(Args::Type::Name),
+            Argument::ForType(Args::Type::Moniker),
+            Argument::ForType(Args::Type::Tag),
+            Argument::ForType(Args::Type::Command),
+            Argument::ForType(Args::Type::Exact),
+            Argument{ "version"_liv, 'v', Args::Type::GatedVersion, Resource::String::GatedVersionArgumentDescription, ArgumentType::Standard },
+            Argument::ForType(Args::Type::Source),
+            Argument::ForType(Args::Type::CustomHeader),
+            Argument::ForType(Args::Type::AcceptSourceAgreements),
+            Argument::ForType(Args::Type::Force),
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Commands/PinCommand.cpp:63]: Curious what is the scenario for --force in pin add?

File: src/AppInstallerCLICore/Commands/PinCommand.cpp [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+            Argument::ForType(Args::Type::Id),
+            Argument::ForType(Args::Type::Name),
+            Argument::ForType(Args::Type::Moniker),
+            Argument::ForType(Args::Type::Source),
+            Argument::ForType(Args::Type::Tag),
+            Argument::ForType(Args::Type::Command),
+            Argument::ForType(Args::Type::Exact),
+            Argument::ForType(Args::Type::CustomHeader),
+            Argument::ForType(Args::Type::AcceptSourceAgreements),
+            Argument::ForType(Args::Type::Force),
```
Comment (yao-msft) [inline src/AppInstallerCLICore/Commands/PinCommand.cpp:126]: Also curious what is the scenario for --force in pin remove here?

**Review by yao-msft — APPROVED** [review]

File: src/AppInstallerCLICore/Commands/PinCommand.cpp (hunk 10 above)
Comment (florelis) [reply src/AppInstallerCLICore/Commands/PinCommand.cpp:63]: The scenario I was thinking of is overwriting the pin. Like `winget pin add Pkg` adds a pin, then `winget pin add Pkg --version 1.*` would fail because it exists, but if you add `--force`, it overwrites it. But honestly, I just added it because it was in the spec and I was hoping you had a scenario in mind :D

Comment (yao-msft) [reply src/AppInstallerCLICore/Commands/PinCommand.cpp:63]: That makes sense. I guess I thought about it some time ago and now my memory does not serve me well.

Comment (Trenly) [reply src/AppInstallerCLICore/Commands/PinCommand.cpp:63]: Even more curious - should `--force` be added to the list of generic arguments (like `--wait` and `--verbose-logs`)? In the worst case, it doesn't change the behavior, and in most cases it will remove a few lines of code.
